src/rust/bitbox02-noise/src/noise_xx.rs

```diff
@@ -0,0 +1,297 @@
+        /// Fetch with `get_handshake_hash()`, used to display the pairing code to verify the
+        /// pairing.
+        handshake_hash: HandshakeHash,
+        /// Communication partner's static public key. Can be used to remember the communication
+        /// partner, so the pairing verification can be skipped the next time.
+        remote_static_pubkey: PublicKey,
+        /// To encrypt outgoing messages.
+        send: noise_protocol::CipherState<ChaCha20Poly1305>,
+        /// To decrypt incoming messages.
+        receive: noise_protocol::CipherState<ChaCha20Poly1305>,
+    },
+}
+
+/// See documentation of `State.handshake()`.
+pub enum HandshakeResult {
+    Response(Vec<u8>),
+    Done,
+}
+
+/// Common error returned by all noise state functions.
+#[derive(Debug)]
+pub enum Error {
+    /// Cannot use this until `set_pairing_verified()` has been called.
+    PairingVerificationRequired,
+    /// Unexpected/internal errors returned by the `noise_protocol` crate.
+    Noise,
+    /// This function was called at the wrong time (see `State` documentation to see in which order
+    /// the functions need be called).
+    WrongState,
+}
+
+impl core::convert::From<Error> for () {
+    fn from(_error: Error) -> Self {}
+}
+
+impl core::convert::From<noise_protocol::Error> for Error {
+    fn from(_error: noise_protocol::Error) -> Self {
+        Error::Noise
+    }
+}
+
+impl<R: Random32> State<R> {
+    /// Can be called at any time to reset the state.
+    pub fn reset(&mut self) {
+        *self = State::Nothing;
+    }
+
+    /// Can be called at any time to start waiting for a new communication channel.
+    ///
+    /// `static_private_key` is the local static key. It can be generated using
+    /// `generate_static_private_key()`.
+    pub fn init(&mut self, static_private_key: Sensitive<PrivateKey>) {
+        let hs = HandshakeState::new(
+            noise_protocol::patterns::noise_xx().clone(),
+            false,
+            &b"Noise_XX_25519_ChaChaPoly_SHA256"[..],
+            Some(static_private_key),
+            None,
+            None,
+            None,
+        );
+        *self = State::Initialized(hs);
+    }
+
+    /// In `noise_XX`, there are 3 handshake messages in total (2 by remote, 1 by us):
+    ///
+    /// `reqA(remote -> local); respB(local -> remote); reqC(remote -> local)`
+    ///
+    /// This function needs to be called twice (once for `A/B`, once for `C`).
+    ///
+    /// See also: [noiseexplorer.com/patterns/XX](https://noiseexplorer.com/patterns/XX/).
+    ///
+    /// Returns: `Response(<respB>)` (see above) for the 1st handshake message, and `Done` for the
+    /// 2nd handshake message.
+    pub fn handshake(&mut self, msg: &[u8]) -> Result<HandshakeResult, Error> {
+        use core::convert::TryInto;
+        match self {
+            State::Initialized(handshake_state) => {
+                let payload = handshake_state.read_message_vec(msg)?;
```

Comment:
It can also fail just due to an invalid message be the app? 

Invalid noise state can happen at anytime, e.g. if one of the sides misses a a round or sth like that.

The app can always reset the noise state by starting a new nosie channel by sending `OP_I_CAN_HAS_HANDSHAEK`.

Reply:
On the other hand probably it won't hurt to reset just in case.

Reply:
Done